test/e2e/autoscale_hpa_test.go

```diff
@@ -0,0 +1,205 @@
+	"knative.dev/serving/pkg/apis/autoscaling"
+	resourcenames "knative.dev/serving/pkg/reconciler/revision/resources/names"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	v1test "knative.dev/serving/test/v1"
+)
+
+const (
+	cpuTarget             = 75
+	targetPods            = 5
+	scaleUpTimeout        = 3 * time.Minute
+	scaleToMinimumTimeout = 10 * time.Minute // 5 minutes is the default window for hpa to calculate if should scale down
+	minimumReplicas       = 1
+	minPods               = 1.0
+	maxPods               = 10.0
+	primeNum              = 500000
+)
+
+func TestHPAAutoscaleUpDownUp(t *testing.T) {
+	ctx := setupHPASvc(t, autoscaling.HPA, autoscaling.CPU, cpuTarget)
+	test.EnsureTearDown(t, ctx.Clients(), ctx.Names())
+	assertHPAAutoscaleUpToNumPods(ctx, minimumReplicas, targetPods, time.After(scaleUpTimeout), true /* quick */)
+	assertScaleDownToOne(ctx)
+	assertHPAAutoscaleUpToNumPods(ctx, minimumReplicas, targetPods, time.After(scaleUpTimeout), true /* quick */)
+}
+
+func setupHPASvc(t *testing.T, class, metric string, target int) *TestContext {
+	t.Helper()
+	clients := Setup(t)
+
+	t.Log("Creating a new Route and Configuration")
+	names := &test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   autoscaleTestImageName,
+	}
+	resources, err := v1test.CreateServiceReady(t, clients, names,
+		[]rtesting.ServiceOption{
+			rtesting.WithConfigAnnotations(map[string]string{
+				autoscaling.ClassAnnotationKey:    class,
+				autoscaling.MetricAnnotationKey:   metric,
+				autoscaling.TargetAnnotationKey:   strconv.Itoa(target),
+				autoscaling.MinScaleAnnotationKey: fmt.Sprintf("%d", minimumReplicas),
+				autoscaling.MaxScaleAnnotationKey: fmt.Sprintf("%d", int(maxPods)),
+			}), rtesting.WithResourceRequirements(corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceCPU:    resource.MustParse("30m"),
+					corev1.ResourceMemory: resource.MustParse("20Mi"),
+				},
+			}),
+		}...)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	if _, err := pkgTest.WaitForEndpointState(
+		context.Background(),
+		clients.KubeClient,
+		t.Logf,
+		names.URL,
+		v1test.RetryingRouteInconsistency(spoof.MatchesAllOf(spoof.IsStatusOK)),
+		"CheckingEndpointAfterCreate",
+		false,
+		test.AddRootCAtoTransport(context.Background(), t.Logf, clients, test.ServingFlags.HTTPS),
+	); err != nil {
+		t.Fatalf("Error probing %s: %v", names.URL.Hostname(), err)
+	}
+
+	return &TestContext{
+		t:           t,
+		logf:        t.Logf,
+		clients:     clients,
+		names:       names,
+		resources:   resources,
+		targetValue: target,
+		metric:      metric,
+	}
+}
+func assertHPAAutoscaleUpToNumPods(ctx *TestContext, curPods, targetPods float64, done <-chan time.Time, quick bool) {
+	ctx.t.Helper()
+	wait := autoscaleHPAUpToNumPods(ctx, curPods, targetPods, done, quick)
+	if err := wait(); err != nil {
+		ctx.t.Fatal(err)
+	}
+}
```

Comment:
This intermediary functions feels unnecessary, unless we want to run this in another test (like the upgrade suite), which I don't think we want currently.

Reply:
Yes I was not sure about it since the kpa tests have it, I can remove.